ToolBox/CoinApi/DefaultConnectionHandler.cs

```diff
@@ -0,0 +1,196 @@
+                            TimeSpan elapsed;
+                            lock (_lockerConnectionMonitor)
+                            {
+                                elapsed = DateTime.UtcNow - _lastDataReceivedTime;
+                            }
+
+                            if (!_connectionLost && elapsed > _maximumIdleTimeSpan)
+                            {
+                                _connectionLost = true;
+                                nextReconnectionAttemptUtcTime = DateTime.UtcNow.AddSeconds(nextReconnectionAttemptSeconds);
+
+                                OnConnectionLost();
+                            }
+                            else if (_connectionLost)
+                            {
+                                if (elapsed <= _maximumIdleTimeSpan)
+                                {
+                                    _connectionLost = false;
+                                    nextReconnectionAttemptSeconds = _minimumSecondsForNextReconnectionAttempt;
+
+                                    OnConnectionRestored();
+                                }
+                                else
+                                {
+                                    if (DateTime.UtcNow > nextReconnectionAttemptUtcTime)
+                                    {
+                                        // double the interval between attempts (capped to 1 minute)
+                                        nextReconnectionAttemptSeconds = Math.Min(nextReconnectionAttemptSeconds * 2, _maximumSecondsForNextReconnectionAttempt);
+                                        nextReconnectionAttemptUtcTime = DateTime.UtcNow.AddSeconds(nextReconnectionAttemptSeconds);
+
+                                        OnReconnectRequested();
+                                    }
+                                }
+                            }
+                        }
+                        catch (Exception exception)
+                        {
+                            Log.Error($"Error in DefaultConnectionHandler: {exception}");
+                        }
+                    }
+                }
+                catch (Exception exception)
+                {
+                    Log.Error(exception);
+                }
+            }) { IsBackground = true };
+
+            _connectionMonitorThread.Start();
+
+            waitHandle.WaitOne();
+        }
+
+        /// <summary>
+        /// Enables/disables monitoring of the connection
+        /// </summary>
+        public void EnableMonitoring(bool isEnabled)
+        {
+            if (_isEnabled) return;
+
+            _isEnabled = isEnabled;
+
+            KeepAlive();
+        }
+
+        /// <summary>
+        /// Notifies the connection handler that new data was received
+        /// </summary>
+        public void KeepAlive()
+        {
+            lock (_lockerConnectionMonitor)
+            {
+                _lastDataReceivedTime = DateTime.UtcNow;
```

Comment:
Being picky about performance here: doing `DateTime.UtcNow` on each new data point might add a significant overhead. Idea: what if we set a flag/event here and have a timer or maybe the `_connectionMonitorThread` check on it every `_maximumIdleTimeSpan`: if not set-> connection lost, else clear flag

Reply:
Good point! 
We could also remove the `DateTime.UtcNow` call by passing in the UTC timestamp of the new tick received, thoughts?

Reply:
Yes, I think that should do the trick too! 👍 it wouldn't require a timezone conversion right? Those are terribly expensive 😄

Reply:
Right, raw data timestamps are UTC for CoinAPI -- should be the same for all other live data feeds 😄